core/src/main/java/org/polypheny/db/catalog/entity/logical/LogicalForeignKey.java

```diff
@@ -0,0 +1,188 @@
+import java.util.List;
+import lombok.EqualsAndHashCode;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.Value;
+import org.polypheny.db.catalog.Catalog;
+import org.polypheny.db.catalog.entity.PolyObject;
+import org.polypheny.db.catalog.logistic.ForeignKeyOption;
+import org.polypheny.db.catalog.snapshot.Snapshot;
+import org.polypheny.db.type.entity.PolyString;
+import org.polypheny.db.type.entity.PolyValue;
+import org.polypheny.db.type.entity.numerical.PolyInteger;
+
+@Value
+@EqualsAndHashCode(callSuper = true)
+public class LogicalForeignKey extends LogicalKey {
+
+    @Serialize
+    public String name;
+
+    @Serialize
+    public long referencedKeyId;
+
+    @Serialize
+    public long referencedKeySchemaId;
+
+    @Serialize
+    public long referencedKeyTableId;
+
+    @Serialize
+    public ForeignKeyOption updateRule;
+
+    @Serialize
+    public ForeignKeyOption deleteRule;
+
+    @Serialize
+    public ImmutableList<Long> referencedKeyColumnIds;
+
+
+    public LogicalForeignKey(
+            @Deserialize("id") final long id,
+            @Deserialize("name") @NonNull final String name,
+            @Deserialize("entityId") final long entityId,
+            @Deserialize("namespaceId") final long namespaceId,
+            @Deserialize("referencedKeyId") final long referencedKeyId,
+            @Deserialize("referencedKeyTableId") final long referencedKeyTableId,
+            @Deserialize("referencedKeySchemaId") final long referencedKeySchemaId,
+            @Deserialize("columnIds") final List<Long> columnIds,
+            @Deserialize("referencedKeyColumnIds") final List<Long> referencedKeyColumnIds,
+            @Deserialize("updateRule") final ForeignKeyOption updateRule,
+            @Deserialize("deleteRule") final ForeignKeyOption deleteRule ) {
+        super( id, entityId, namespaceId, columnIds, EnforcementTime.ON_COMMIT );
+        this.name = name;
+        this.referencedKeyId = referencedKeyId;
+        this.referencedKeyTableId = referencedKeyTableId;
+        this.referencedKeySchemaId = referencedKeySchemaId;
+        this.referencedKeyColumnIds = ImmutableList.copyOf( referencedKeyColumnIds );
+        this.updateRule = updateRule;
+        this.deleteRule = deleteRule;
+    }
+
+
+    public String getReferencedKeySchemaName() {
+        return Catalog.snapshot().getNamespace( referencedKeySchemaId ).orElseThrow().name;
+    }
+
+
+    public String getReferencedKeyTableName() {
+        return Catalog.snapshot().rel().getTable( referencedKeyTableId ).orElseThrow().name;
+    }
+
+
+    public List<String> getReferencedKeyColumnNames() {
+        Snapshot snapshot = Catalog.snapshot();
+        List<String> columnNames = new LinkedList<>();
+        for ( long columnId : referencedKeyColumnIds ) {
+            columnNames.add( snapshot.rel().getColumn( columnId ).orElseThrow().name );
+        }
+        return columnNames;
+    }
```

Comment:
Currently, we only support foreign keys in the relational schema. However, conceptually can also support this for the other data models as well.

Reply:
Definitely, we should also have this not as relational as it is now but use generic names and methods, as this is not a strictly relational concept. I refactored it as much as it should make sense for now.